lib/internal/modules/esm/resolve.js

```diff
@@ -377,13 +376,6 @@ function finalizeResolution(resolved, base, preserveSymlinks) {
 
   let path = fileURLToPath(resolved);
   if (getOptionValue('--experimental-specifier-resolution') === 'node') {
-    if (!experimentalSpecifierResolutionWarned) {
-      process.emitWarning(
-        'The Node.js specifier resolution flag is experimental. It could change or be removed at any time.',
```

Comment:
It would appear that this PR removes the actual text of the warning and replaces it with placeholders? We want to preserve the current messages.

Reply:
[`emitExperimentalWarning()`](https://github.com/nodejs/node/blob/457567f72ca73ef000b56d30f1f6b60e251cc15a/lib/internal/util.js#L219-L225) has almost identical wording (I think it's just missing the "The Node.js" at the beginning)

Reply:
It also has the added benefit that it will not emit the same warning more than once (based on the feature name supplied).

Reply:
What gets printed? We just changed this a few days ago in #42252.

Reply:
#42252 is included in this PR :)

[42252's wording](https://github.com/nodejs/node/pull/42252/files#diff-b4c24f634e3741e5ad9e8c29864a48f2bd284a9d66d8fed3d077ccee0c44087bR380-R383):

> The Node.js specifier resolution flag is experimental. It could change or be removed at any time.

In this PR:

> ExperimentalWarning: Specifier Resolution is an experimental feature. This feature could change at any time

And the other two affected by this PR:

> ExperimentalWarning: Custom ESM Loaders is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)

> ExperimentalWarning: Network Imports is an experimental feature. This feature could change at any time

If the current "Node.js" part before "specifier resolution" is significant, I'm happy to add that (I wasn't sure if it was signifying the feature was Node.js, or if it was referring to the "node" value passed to the flag).

Reply:
For specifier resolution, I want to get across that the flag could not just change but also be removed. Could we put that one back to the custom wording?

Reply:
Ahh, sorry. Yes. I shall do (I'll have to switch back to `process.emitWarning()` and add a little flag to avoid a redundant warning).

Reply:
Soo this actually breaks the heck outta the test—for some reason, it never gets passed the internal instance of the ESMLoader, which I'm (now) using as a gate to avoid a redundant print for specifier resolution.

```js
ESMLoader {
  constructor(isInternal = false) {
    if (isInternal) return;

    if (getOptionValue('--experimental-loader')) {
      emitExperimentalWarning('Custom ESM Loaders');
    }
    if (getOptionValue('--experimental-network-imports')) {
      emitExperimentalWarning('Network Imports');
    }
    if (getOptionValue('--experimental-specifier-resolution') === 'node') {
      process.emitWarning(
        'The Node.js specifier resolution flag is experimental. It could change or be removed at any time.',
        'ExperimentalWarning'
      );
    }
  }
}
```

Reply:
> Soo this actually breaks the heck outta the test—for some reason, it never gets passed the internal instance of the ESMLoader, which I’m (now) using as a gate to avoid a redundant print for specifier resolution.

Maybe use both? Something like this:

```js
    if (getOptionValue('--experimental-specifier-resolution') === 'node') {
      emitExperimentalWarning('The Node.js specifier resolution flag');
      process.emitWarning(
        'The Node.js specifier resolution flag will be removed once the Loaders API provides equivalent functionality.',
        'ExperimentalWarning'
      );
```

I’d think that would generate output like this:

```
ExperimentalWarning: The Node.js specifier resolution flag is an experimental feature. This feature could change at any time
ExperimentalWarning: The Node.js specifier resolution flag will be removed once the Loaders API provides equivalent functionality.
```

I don’t know why “This feature could change at any time” lacks a period, but it should have one.

Anyway maybe this is overkill and/or ugly, in which case we should just get the `process.emitWarning` version to pass a test. Perhaps if you wrote a test that used a child process, like https://github.com/nodejs/node/commit/a30e0de0059f7d49e1adbbc2f8197b08069a20a6#diff-709c4b77b4c83e12adaaa1ab856625287a7c2b1b56d1836d179afef22ab41435, that would solve the issue?

Reply:
If I used both but without the `isInternal` gate, you'd get:

> ExperimentalWarning: The Node.js specifier resolution flag is an experimental feature. This feature could change at any time
> ExperimentalWarning: The Node.js specifier resolution flag will be removed once the Loaders API provides equivalent functionality.
> ExperimentalWarning: The Node.js specifier resolution flag will be removed once the Loaders API provides equivalent functionality.

(The custom message printed twice)

I am using that child process test. On child process close, only the internal ESMLoader has been called.

IRL, it happens as expected 😵‍💫

Yes, I noticed the lack of a period from `emitExperimentalWarning()`. I was gonna fix it in a separate PR.

Reply:
@GeoffreyBooth it seems the presence of `process.emitWarning()` causes the test to fail, regardless of the sequence of the cases. I think it's maybe causing the test to terminate early, but I don't know why. All I can think of is it causes the child process to close before it should. With `process.emitWarning()` instead of `emitExperimentalWarning()`, the test does not behave the same as the real world, which is troubling.

Reply:
Well I think we should fix this before merging in this PR. Do you want to get this branch in the ideal state, where the code and test are what you think each should be if only the test would pass? And then I or someone could try to figure out how to get the test to pass.

Reply:
I have that currently; shall I push with the test broken?

Reply:
I think so. I'm not sure when I'll be able to take a look, so feel free to go ahead rather than wait for me.

Reply:
pushed. with your requested change, it can't merge though.